R/helper.R

```diff
@@ -19,3 +19,39 @@ get_by_id = function(xs, id) {
   }
   return(NULL)
 }
+
+
+#' @description calculate pareto front of an all-numeric data frame (or matrix)
+#'   also works for >2D pareto front
+#' @example data_frame = data.frame(c(1, 0.25, 0.75, 0), c(0, 0.25, 0.75, 1))
+#'   calculate_pareto_front(data_frame)
+#'   data_frame = data.frame(c(1,1,0,1), c(1,1,0,0), c(1,0,1,0), c(1,1,1,1))
+#'   calculate_pareto_front(data_frame, maximize = FALSE)
+calculate_pareto_front = function(data_frame, maximize = TRUE) {
+
+  assert_data_frame(data_frame, types = "numeric")
+  assert_logical(maximize)
+
+  # sort the whole data.frame and go stepwise through the columns whenever
+  # entries are tied
+  sorted = do.call(order, c(as.list(data_frame), decreasing = maximize))
```

Comment:
In case of real ties (tied in all columns), we pick the first of the tied rows:
```{r}
  data_frame = data.frame(x = c(1, 1), y = c(0, 0))
  calculate_pareto_front(data_frame)
```

This is rarely a good idea. Either return all, or select one randomly. For the latter, pass `rnorm(nrow(data_frame))` as last vector to `order()`:
```r
do.call(order, c(as.list(data_frame), runif(nrow(data_frame)), list(decreasing = maximize))
```